packages/evm/test/core/signature-validator.js

```diff
@@ -0,0 +1,611 @@
+        assertBn(nextNonce, user === sender ? previousNonce : previousNonce.add(bn(1)), 'next nonce does not match')
+      })
+    }
+
+    const encodeCalldata = async (extraCalldata = undefined) => {
+      const ABI = validator.abi.find(i => i.name === 'authenticateCall')
+      const calldata = abi.encodeFunctionCall(ABI, [user])
+      return extraCalldata === undefined ? encodeExtraCalldata(calldata) : `${calldata}${extraCalldata}`
+    }
+
+    const encodeExtraCalldata = async (calldata) => {
+      const domainSeparator = await validator.getDomainSeparator()
+      const ABI = validator.abi.find(i => i.name === allowedFunctionality)
+      const allowedData = abi.encodeFunctionCall(ABI, [user])
+      const encodedData = keccak256(abi.encodeParameters(['bytes', 'address', 'uint256', 'uint256'], [allowedData, allowedSender, nonce, deadline]))
+
+      const digest = soliditySha3(
+        { type: 'bytes1', value: '0x19' },
+        { type: 'bytes1', value: '0x01' },
+        { type: 'bytes32', value: domainSeparator },
+        { type: 'bytes32', value: encodedData }
+      )
+
+      const signature = sign(digest)
+      return encodeSignature(calldata, deadline, signature)
+    }
+
+    const setUser = address => {
+      beforeEach(`set user ${address}`, async () => {
+        user = address
+      })
+    }
+
+    const setAuthorizedSender = address => {
+      beforeEach(`set authorized sender ${address}`, async () => {
+        allowedSender = address
+      })
+    }
+
+    const setAuthorizedFunctionality = fnName => {
+      beforeEach(`set authorized functionality ${fnName}`, async () => {
+        allowedFunctionality = fnName
+      })
+    }
+
+    const setNonce = (offset = 0) => {
+      beforeEach(`set nonce with offset ${offset}`, async () => {
+        const nextNonce = await validator.getNextNonce(user)
+        nonce = nextNonce.add(bn(offset))
+      })
+    }
+
+    const setDeadline = (offset = 0) => {
+      beforeEach(`set deadline with offset ${offset}`, async () => {
+        const currentTimestamp = await validator.getTimestampExt()
+        deadline = currentTimestamp.add(bn(offset))
+      })
+    }
+
+    context('when there is no extra calldata given', () => {
+      const extraCalldata = ''
+
+      context('when there sender and the user are the same', () => {
+        setUser(sender)
+        itAllowsTheSender(extraCalldata)
+      })
+
+      context('when there sender and the user are different accounts', () => {
+        setUser(externalAccount)
+        itReverts(extraCalldata)
+      })
+    })
+
+    context('when there is some extra calldata given', () => {
+      context('when the sender and the user are the same', () => {
+        setUser(sender)
+
+        context('when the extra calldata is malformed', () => {
+          const extraCalldata = 'abcd'
+
+          itAllowsTheSender(extraCalldata)
+        })
+
+        context('when the extra calldata is well formed', () => {
```

Comment:
Just clarifying, here when we set bad variables (e.g. `setDeadline(-100)`), it doesn't matter because the extra calldata isn't checked?

Reply:
The `setDeadline` argument is actually an offset for the current timestamp

Reply:
Right, that's why I wasn't sure. Using `setDeadline(-100)` (and likewise, other bad constructions in this context) should fail, but they're expected to pass.